src/lib/ams.c

```diff
@@ -0,0 +1,749 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright (c) 2022 Intel Corporation
+//
+// Author: Krzysztof Frydryk <krzysztofx.frydryk@intel.com>
+
+#include <sof/lib/memory.h>
+#include <sof/lib/cpu.h>
+#include <sof/coherent.h>
+#include <sof/schedule/ll_schedule.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/schedule/schedule.h>
+#include <sof/ipc/topology.h>
+#include <sof/drivers/idc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <sof/lib/ams.h>
+
+LOG_MODULE_REGISTER(ams, CONFIG_SOF_LOG_LEVEL);
+
+DECLARE_SOF_UUID("ams", ams_uuid, 0xea9c4bca, 0x5b7d, 0x48c6,
+		 0x95, 0x86, 0x55, 0x3e, 0x27, 0x23, 0x5b, 0xeb);
+
+DECLARE_TR_CTX(ams_tr, SOF_UUID(ams_uuid), LOG_LEVEL_INFO);
+
+/* AMS routing table */
+const uint8_t g_ams_core_routing[CONFIG_CORE_COUNT][CONFIG_CORE_COUNT] = {
+/* to core: 0..N */
+/* from core 1 */ {
+	0,
+	1,
+	2,
+	}, /* master core can target every other core directly */
+/* from core 2 */ {
+	0,
+	1,
+	0,
+	}, /* and slaves must always go thru master core */
+/* from core 3 */ {
+	0,
+	0,
+	2,
+}};
+
+static struct am_service_context ctx;
+static struct am_service_shared_context shared_ctx;
+static struct ams_routing_entry rt_table[AM_ROUTING_TAB_SIZE];
+static struct ams_producer producer_table[AM_ROUTING_TAB_SIZE];
+static struct uuid_idx uuid_table[AM_SERVICE_UUID_TABLE_SIZE];
```

Comment:
would be good to unify these macros: maybe use a common `AMS_` prefix? And `TABLE` for all? Same for types and function names - should they all begin with ams_?